src/oc/storage/api/boards.clj

```diff
@@ -0,0 +1,290 @@
+            [oc.lib.schema :as lib-schema]
+            [oc.lib.slugify :as slugify]
+            [oc.lib.db.pool :as pool]
+            [oc.lib.api.common :as api-common]
+            [oc.storage.config :as config]
+            [oc.storage.api.access :as access]
+            [oc.storage.representations.media-types :as mt]
+            [oc.storage.representations.board :as board-rep]
+            [oc.storage.representations.entry :as entry-rep]
+            [oc.storage.resources.common :as common-res]
+            [oc.storage.resources.org :as org-res]
+            [oc.storage.resources.board :as board-res]
+            [oc.storage.resources.entry :as entry-res]))
+
+;; ----- Utility functions -----
+
+(defn- assemble-board [conn org-slug slug ctx]
+  (let [board (or (:updated-board ctx) (:existing-board ctx))
+        topic-slugs (map name (:topics board)) ; slug for each active topic
+        entries (entry-res/get-entries-by-board conn (:uuid board)) ; latest entry for each topic
+        selected-entries (select-keys entries topic-slugs) ; active entries
+        selected-entry-reps (zipmap topic-slugs
+                              (map #(entry-rep/render-entry-for-collection org-slug slug
+                                      (get selected-entries %) (:access-level ctx))
+                                topic-slugs))
+                                 archived-entries (clojure.set/difference (set (keys entries)) (set topic-slugs))
+                                 archived (map #(identity {:slug % :title (:title (get entries %))}) archived-entries)
+        authors (:authors board)
+        author-reps (map #(board-rep/render-author-for-collection org-slug slug %) authors)
+        viewers (:viewers board)
+        viewer-reps (map #(board-rep/render-viewer-for-collection org-slug slug %) viewers)]
+    (merge (-> board 
+              (assoc :archived archived)
+              (assoc :authors author-reps)
+              (assoc :viewers viewer-reps))
+      selected-entry-reps)))
+
+;; ----- Validations -----
+
+(defn- valid-new-board? [conn org-slug {board-map :data author :user}]
+  (if-let [org (org-res/get-org conn org-slug)]
+    (try
+      {:new-board (board-res/->board (:uuid org) board-map author) :existing-org org}
+
+      (catch clojure.lang.ExceptionInfo e
+        [false, {:reason (.getMessage e)}])) ; Not a valid new board
+    [false, {:reason :invalid-org}])) ; couldn't find the specified org
+
+(defn- valid-board-update? [conn org-slug slug board-props]
+  (if-let* [org (org-res/get-org conn org-slug)
+            board (board-res/get-board conn (:uuid org) slug)]
+    (let [updated-board (merge board (board-res/clean board-props))]
+      (if (lib-schema/valid? common-res/Board updated-board)
+        {:existing-org org :existing-board board :board-update updated-board}
+        [false, {:board-update updated-board}])) ; invalid update
+    true)) ; No org or board, so this will fail existence check later
```

Comment:
The existence checks that happen later will be more specific.

I don't love this part of Liberator that the input validity checks happen before the existence checks. Seems backwards to me.